### PR TITLE
Tribal Section 2 Parsing & Validation

### LINKED ISSUE #1115: Tribal TANF Closed Data (02) File Parsing - As a data analyst, I want my data files to be parsed so that my submission can be validated/accepted by TDP
**Description:**
_Provide a brief background and justification for this issue_
Upon file upload, we expect TDP to attempt parsing per the spec outlined:
https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_data_report_section2.pdf

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] Allow Tribes to submit Closed Data based on the TANF data parsing spec

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 
- [ ] Ensuring file content type is tribal TANF. 
- [ ] Correctly formatted input data can be built into a Python data structure
- [ ] Completed Python data structure can be translated to database model(s)
- [ ] Incorrectly formatted data is identified and mistake(s) are logged.
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 
This is the same for TANF section 2

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Added support for Tribal TANF section 2 parsing and validation
Pull request closes #1115 

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
- Submit both `ADS.E2J.FTP2.TS142.txt` 
- Verify the file generates Tribal T4/T5 records.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Allow Tribes to submit Closed Data based on the TANF data parsing spec

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Review by andrew-jameson — APPROVED** [review]
LGTM

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    schemas=[
+        RowSchema(
+            model=Tribal_TANF_T4,
+            preparsing_validators=[
+                validators.hasLength(71),
+            ],
+            postparsing_validators=[],
+            fields=[
+                Field(
+                    item="0",
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py:20]: Please add friendly_name for each field as per #2599

**Review by raftmsohani — DISMISSED** [review]
Please add friendly_name for each field. for ref, look at 2599

**Review by jtimpe — APPROVED** [review]
Agree with Mo about adding a `friendly_name`, otherwise looks good!

**Review by George-Hudson — APPROVED** [review]
LGTM

**Comment by elipe17** [comment]
@ADPennington, Updated based on your feedback and got it deployed to QASP.

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+                    validators=[validators.validateRace()],
+                ),
+                Field(
+                    item="17C",
+                    name="RACE_ASIAN",
+                    friendly_name="race asian",
+                    type="number",
+                    startIndex=39,
+                    endIndex=40,
+                    required=True,
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py]:
```suggestion
                    required=False,
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+                    validators=[validators.isInLimits(0, 2)],
+                ),
+                Field(
+                    item="19B",
+                    name="REC_FEDERAL_DISABILITY",
+                    friendly_name="receives federal disability",
+                    type="number",
+                    startIndex=45,
+                    endIndex=46,
+                    required=True,
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py]:
```suggestion
                    required=False,
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+                    validators=[validators.isInLimits(1, 2)],
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py]:
```suggestion
                    validators=[validators.isInLimits(0, 2)],
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+                    validators=[validators.isInLimits(0, 2)],
+                ),
+                Field(
+                    item="19E",
+                    name="REC_SSI",
+                    friendly_name="receives social security income",
+                    type="number",
+                    startIndex=48,
+                    endIndex=49,
+                    required=True,
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py]:
```suggestion
                    required=False,
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+                    validators=[validators.isInLimits(1, 2)],
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py]:
```suggestion
                    validators=[validators.isInLimits(0, 2)],
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+                    ),
+                ),
+                validators.if_then_validator(
+                    condition_field="FAMILY_AFFILIATION",
+                    condition_function=validators.matches(1),
+                    result_field="CITIZENSHIP_STATUS",
+                    result_function=validators.isInLimits(1, 2),
+                ),
+                validators.validate__FAM_AFF__HOH__Count_Fed_Time(),
+                validators.if_then_validator(
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t5.py:91]: @elipe17 this can be dropped. we cant validate this without calculating age.

**Review by ADPennington — APPROVED** [review]
@elipe17 this is in good shape. a few suggestions/comments on validators included below. once these are addressed, please re-deploy :)

**Review by ADPennington — CHANGES_REQUESTED** [review]
changes requested.

**Comment by elipe17** [comment]
@ADPennington deploying now

**Review by ADPennington — APPROVED** [review]
🚀

File: tdrs-backend/tdpservice/search_indexes/test/test_model_mapping.py [hunk 8]
```diff
@@ -902,3 +902,85 @@ def test_can_create_and_index_tribal_tanf_t3_submission(test_datafile):
...
+
+    assert submission.id is not None
+
+    search = documents.tribal.Tribal_TANF_T5DataSubmissionDocument.search().query(
+        'match',
+        RecordType=record_num
+    )
+    response = search.execute()
+
+    assert response.hits.total.value == 1
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/test/test_model_mapping.py:986]: can we assert specific fields instead?

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
+"""Schema for Tribal TANF T4 record types."""
+
+from ...util import SchemaManager
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t4.py:3]: Note: I have moved the SchemaManager in my PR, depending on who merges first we might see a conflict

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Pull request merged**